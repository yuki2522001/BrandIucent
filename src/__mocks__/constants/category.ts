import { Categories } from "@common-types/category";
import url from "@assets/images/icons/beverages.png";

const CATEGORY: Categories = {
  id: "1234",
  name: "Beverages",
  images: {
    src: url,
    alt: "This is icon beverages",
  },
};

const CATEGORY_MOCKING_LIST: Categories[] = [
  {
    id: "1234",
    name: "Beverages",
    images: {
      src: "@assets/images/icons/beverages.png",
      alt: "This is icon beverages",
    },
  },
  {
    id: "12345",
    name: "Dairy",
    images: {
      src: "@assets/images/icons/dairy.png",
      alt: "This is icon dairy",
    },
  },
  {
    id: "12345",
    name: "Meat",
    images: {
      src: "@assets/images/icons/meat.png",
      alt: "This is icon meat",
    },
  },
  {
    id: "12345",
    name: "Spices & Herbs",
    images: {
      src: "@assets/images/icons/spices-herbs.png",
      alt: "This is icon spices-herbs",
    },
  },
  {
    id: "12345",
    name: "Stationery",
    images: {
      src: "@assets/images/icons/stationery.png",
      alt: "This is icon stationery",
    },
  },
  {
    id: "12345",
    name: "Costmetic",
    images: {
      src: "@assets/images/icons/costmetic.png",
      alt: "This is icon costmetic",
    },
  },
];

export { CATEGORY, CATEGORY_MOCKING_LIST };
