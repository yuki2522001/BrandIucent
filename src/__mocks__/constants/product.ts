import { Product } from "@common-types/product";
import iceCreamSundae from "@assets/images/ice-cream-sundae.png";
import beefSteak from "@assets/images/beef-steak.png";

const PRODUCT_MOCKING: Product = {
  productId: "16466545464646547",
  categoryId: "1651454657522",
  productName: "Ice Cream Sundae",
  productDescription:
    "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam molestie ultrices donec pulvinar quis est bibendum suspendisse viverra.",
  productQuantity: 5,
  originalPrice: {
    value: 20,
    currency: "$",
  },
  discountPrice: {
    value: 20,
    currency: "$",
  },
  offerQuantity: "3kg",
  offer: false,
  bestSelling: true,
  popular: false,
  images: {
    src: iceCreamSundae,
    alt: "This is ice-cream-sundae images",
  },
};
const PRODUCT_MOCKING_LIST: Product[] = [
  {
    productId: "16466545464646547",
    categoryId: "1651454657522",
    productName: "Ice Cream Sundae",
    productDescription:
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam molestie ultrices donec pulvinar quis est bibendum suspendisse viverra.",
    productQuantity: 5,
    originalPrice: {
      value: 20,
      currency: "$",
    },
    discountPrice: {
      value: 20,
      currency: "$",
    },
    offerQuantity: "3kg",
    offer: false,
    bestSelling: true,
    popular: false,
    images: {
      src: iceCreamSundae,
      alt: "This is ice-cream-sundae images",
    },
  }
];

export { PRODUCT_MOCKING, PRODUCT_MOCKING_LIST };
