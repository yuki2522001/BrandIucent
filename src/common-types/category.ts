export interface Categories {
  id: string;
  name: string;
  images: {
    src: string;
    alt: string;
  };
}