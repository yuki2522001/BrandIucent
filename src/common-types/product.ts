export interface Product {
  productId: string;
  categoryId: string;
  productName: string;
  productDescription: string ;
  productQuantity: number;
  originalPrice: {
    value: number;
    currency: string;
  };
  discountPrice: {
    value: number;
    currency: string;
  };
  offerQuantity: string;
  offer: boolean;
  bestSelling: boolean;
  popular: boolean;
  images: {
    src: string;
    alt: string;
  }
}

export interface ProductProps {
  product?: Product;
}