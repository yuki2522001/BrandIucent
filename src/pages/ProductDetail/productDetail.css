.productDetail {
  position: relative;
  z-index: 15;
}

.productDetail__background {
  position: absolute;
  z-index: 5;
  top: 0;
  right: 0;
  margin-bottom: 70px;
}

.productDetail__image {
  margin: 358px 0 335px 287px;
}

.productDetail__price,
.productDetail__description {
  margin: 28px 0 13px;
}

.productDetail__counter {
  position: relative;
  margin: 34px 0 50px;
}

.productDetail__price,
.productDetail__quantity {
  font-weight: 400;
  font-size: 18px;
  line-height: 27px;
  color: var(--border-primary);
}

.productDetail__description {
  font-weight: 400;
  font-size: 18px;
  line-height: 27px;
  color: var(--text-primary);
  width: 638px;
}

.productDetail__info {
  position: absolute;
  z-index: 10;
  top: 288px;
  left: 970px;
}

.productDetail__btn {
  margin-left: 232px;
}

.productDetail__more {
  padding: 160px 0 0 100px;
}

.productDetail__productMore {
  display: flex;
  flex-wrap: wrap;
}

.productDetail__title {
  margin: 90px 0 44px 0;
}