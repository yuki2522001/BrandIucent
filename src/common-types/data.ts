import { Categories } from "./category";
import { Product } from "./product";

export interface DataContextProps {
  products: Product[];
  setProducts: Function;
  searchValue: [] | undefined;
  setSearchValue: Function;
  categories: Categories[];
  setCategories: Function;
}