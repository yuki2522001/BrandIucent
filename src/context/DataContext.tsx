import { createContext, useMemo, useState } from "react";
import { Categories } from "@common-types/category";
import { DataContextProps } from "@common-types/data";
import { Product } from "@common-types/product";

const DataContext = createContext<DataContextProps>({} as DataContextProps);
const DataProvider: React.FC<{ children: JSX.Element[] | JSX.Element }> = ({
  children,
}) => {
  const [searchValue, setSearchValue] = useState<[]>([]);
  const [categories, setCategories] = useState<Categories[]>([]);
  const [products, setProducts] = useState<Product[]>([]);

  const value = {
    products,
    setProducts,
    searchValue,
    setSearchValue,
    categories,
    setCategories,
  };

  return <DataContext.Provider value={value}>{children}</DataContext.Provider>;
};

export { DataProvider, DataContext };
