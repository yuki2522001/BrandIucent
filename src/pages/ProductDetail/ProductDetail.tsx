import React, { useContext } from "react";
import { DataContext } from "@context/DataContext";
import { useParams } from "react-router-dom";
import NavigationBar from "@components/common/NavigationBar/NavigationBar";
import background from "@assets/images/backgrounds/gray.png";
import Text from "@components/common/Text/Text";
import Price from "@components/common/Price/Price";
import Counter from "@components/common/Counter/Counter";
import Button from "@components/common/Button/Button";
import ProductList from "@components/ProductList/ProductList";
import "./productDetail.css";

const ProductDetail: React.FC = () => {
  const { id } = useParams();
  const { products } = useContext(DataContext);
  const dataElement = products?.find((item) => item.productId === id);
  const productMore = products.filter((product) =>
    dataElement?.categoryId.includes(product.categoryId)
  );

  return (
    <div data-testid="product-detail" className="productDetail">
      <NavigationBar mode="dark" />
      <figure>
        <img
          className="productDetail__image"
          src={dataElement?.images.src}
          alt={dataElement?.images.alt}
        />
        <img
          className="productDetail__background"
          src={background}
          alt="This is background"
        />
      </figure>
      <div className="productDetail__info">
        <Text text={dataElement?.productName} size="large" />
        <p className="productDetail__price">Price</p>
        <Price
          value={dataElement?.originalPrice.value}
          type="original"
          currency={dataElement?.originalPrice.currency}
        />
        <p className="productDetail__description"></p>
        <p className="productDetail__quantity">Quantity</p>
        <div className="productDetail__counter">
          <Counter counter={dataElement?.productQuantity} />
        </div>
        <div className="productDetail__btn">
          <Button text="Add to cart" type="large" />
        </div>
      </div>
      <div className="productDetail__more">
        <div className="productDetail__title">
          <Text text="Products you may like" decoration="dark" />
        </div>
        <div className="productDetail__productMore">
          <ProductList
            type="medium-box"
            content="quantity"
            visibleCounter={true}
            productList={productMore}
          />
        </div>
      </div>
    </div>
  );
};

export default ProductDetail;
