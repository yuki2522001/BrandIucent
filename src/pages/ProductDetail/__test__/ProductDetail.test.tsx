import { render } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import { DataContextProps } from "@common-types/data";
import { CATEGORY_MOCKING_LIST } from "@__mocks__/constants/category";
import { PRODUCT_MOCKING_LIST } from "@__mocks__/constants/product";
import { DataContext } from "@context/DataContext";
import "@testing-library/jest-dom";
import ProductDetail from "../ProductDetail";

const contextProductMock: DataContextProps = {
  searchValue: [],
  setSearchValue: jest.fn(),
  categories: CATEGORY_MOCKING_LIST,
  setCategories: jest.fn(),
  products: PRODUCT_MOCKING_LIST,
  setProducts: jest.fn(),
};

describe("ProductDetail component", () => {
  test("should render ProductDetail component", () => {
    const history = createMemoryHistory();
    const { getByTestId } = render(
      <DataContext.Provider value={contextProductMock}>
        <Router location={history.location} navigator={history}>
          <ProductDetail />
        </Router>
      </DataContext.Provider>
    );
    expect(getByTestId("product-detail")).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <DataContext.Provider value={contextProductMock}>
        <Router location={history.location} navigator={history}>
          <ProductDetail />
        </Router>
      </DataContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
