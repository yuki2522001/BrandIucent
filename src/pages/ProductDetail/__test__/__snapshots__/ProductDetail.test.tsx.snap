// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetail component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="productDetail"
    data-testid="product-detail"
  >
    <nav
      class="nav"
      data-testid="navigation-bar"
    >
      <a
        data-testid="logo"
        href="/"
      >
        <img
          alt="This is logo"
          class=""
        />
      </a>
      <div
        class="nav__about"
      >
        <div
          class="menu"
          data-testid="menu"
        >
          <ul
            class="menu__list"
          >
            <li
              class="menu__item"
              data-testid="menu-item"
            >
              <a
                href="/products"
              >
                Shop
              </a>
            </li>
            <li
              class="menu__item"
            >
              <a
                href="/"
              >
                Orders
              </a>
            </li>
            <li
              class="menu__item"
            >
              <a
                href="/"
              >
                Cart
              </a>
            </li>
            <li
              class="menu__item"
            >
              <a
                href="/"
              >
                Contact
              </a>
            </li>
          </ul>
        </div>
        <button
          class="btn btn--dark"
        >
          Sign Up
        </button>
        <button
          class="btn btn__outline--dark"
        >
          Login
        </button>
      </div>
    </nav>
    <figure>
      <img
        class="productDetail__image"
      />
      <img
        alt="This is background"
        class="productDetail__background"
      />
    </figure>
    <div
      class="productDetail__info"
    >
      <p
        class="text--large"
        data-testid="text"
      />
      <p
        class="productDetail__price"
      >
        Price
      </p>
      <p
        class="price__original"
        data-testid="price"
      >
        <span
          class="price__unit"
        />
      </p>
      <p
        class="productDetail__description"
      />
      <p
        class="productDetail__quantity"
      >
        Quantity
      </p>
      <div
        class="productDetail__counter"
      >
        <div
          class="counter"
          data-testid="counter"
        >
          <div
            class="counter__minus"
          >
            <i
              class="icon-minus"
              data-testid="icon"
            />
          </div>
          <input
            class="counter__input"
            min="0"
            value=""
          />
          <div
            class="counter__plus"
          >
            <i
              class="icon-plus"
              data-testid="icon"
            />
          </div>
        </div>
      </div>
      <div
        class="productDetail__btn"
      >
        <button
          class="btn btn--large"
        >
          Add to cart
        </button>
      </div>
    </div>
    <div
      class="productDetail__more"
    >
      <div
        class="productDetail__title"
      >
        <p
          class="text--dark"
          data-testid="text"
        >
          Products you may like
        </p>
      </div>
      <div
        class="productDetail__productMore"
      />
    </div>
  </div>
</DocumentFragment>
`;
